[DOC] Adding guides to explain UDF serialization and Broadcast variable usage

Niharikadutta opened the pull request:
We are excited to review your PR.

So we can do the best job, please check:

- [x] There's a descriptive title that will make sense to other developers some time from now. 
- [ ] There's associated issues. All PR's should have issue(s) associated - unless a trivial self-evident change such as fixing a typo. You can use the format `Fixes #nnnn` in your description to cause GitHub to automatically close the issue(s) when your PR is merged.
- [ ] Your change description explains what the change does, why you chose your approach, and anything else that reviewers should know.
- [ ] You have included any necessary tests in the same PR.

Niharikadutta commented:
@bamurtaugh Could you please take a look at the following guides I wrote to explain the UDF serialization behavior and Broadcast variable usage? Thank you for your time!

bamurtaugh commented on docs/broadcast-guide.md:
```diff
@@ -0,0 +1,89 @@
+# Guide to using Broadcast Variables
+
+This is a guide to show how to use Broadcast variables in .NET for Apache Spark.
```
nit: Should we capitalize broadcast in the middle of sentences?

Niharikadutta replied:
Good point, I will change that

bamurtaugh commented on docs/broadcast-guide.md:
```diff
@@ -0,0 +1,89 @@
+
+## What are Broadcast variables
+
+[Broadcast variables in Apache Spark](https://spark.apache.org/docs/2.2.0/rdd-programming-guide.html#broadcast-variables) is a mechanism for sharing variables across executors that are meant to be read-only. They allow the programmer to keep a read-only variable cached on each machine rather than shipping a copy of it with tasks. They can be used, for example, to give every node a copy of a large ...
```
```suggestion
[Broadcast variables in Apache Spark](https://spark.apache.org/docs/2.2.0/rdd-programming-guide.html#broadcast-variables) are a mechanism for sharing variables across executors that are meant to be read-only. They allow the programmer to keep a read-only variable cached on each machine rather than shipping a copy of it with tasks. They can be used, for example, to give every node a copy of a large input dataset in an efficient manner.
```

bamurtaugh commented on docs/broadcast-guide.md:
```diff
@@ -0,0 +1,89 @@
+# Guide to using Broadcast Variables
+
+This is a guide to show how to use Broadcast variables in .NET for Apache Spark.
```
```suggestion
This is a guide to show how to use broadcast variables in .NET for Apache Spark.
```

bamurtaugh commented on docs/broadcast-guide.md:
```diff
@@ -0,0 +1,89 @@
+
+This is a guide to show how to use Broadcast variables in .NET for Apache Spark.
+
+## What are Broadcast variables
```
```suggestion
## What are broadcast variables
```

bamurtaugh replied:
Or, "What are Broadcast Variables"

bamurtaugh commented on docs/broadcast-guide.md:
```diff
@@ -0,0 +1,89 @@
+
+[Broadcast variables in Apache Spark](https://spark.apache.org/docs/2.2.0/rdd-programming-guide.html#broadcast-variables) is a mechanism for sharing variables across executors that are meant to be read-only. They allow the programmer to keep a read-only variable cached on each machine rather than shipping a copy of it with tasks. They can be used, for example, to give every node a copy of a large ...
+
+### How to use Broadcast variables in .NET for Apache Spark
```
```suggestion
### How to use broadcast variables in .NET for Apache Spark
```

bamurtaugh commented on docs/broadcast-guide.md:
```diff
@@ -0,0 +1,89 @@
+
+### How to use Broadcast variables in .NET for Apache Spark
+
+Broadcast variables are created from a variable `v` by calling `SparkContext.Broadcast(v)`. The broadcast variable is a wrapper around `v`, and it's value can be accessed by calling the `Value()` method on it. 
```
```suggestion
Broadcast variables are created from a variable `v` by calling `SparkContext.Broadcast(v)`. The broadcast variable is a wrapper around `v`, and its value can be accessed by calling the `Value()` method on it. 
```

bamurtaugh commented on docs/broadcast-guide.md:
````diff
@@ -0,0 +1,89 @@
+	str => $"{str}: {bv.Value()}");
+```
+
+The type of broadcast variable is captured by using Generics in C# as can be seen in the above example.
````
```suggestion
The type of broadcast variable is captured by using Generics in C#, as can be seen in the above example.
```

bamurtaugh commented on docs/broadcast-guide.md:
````diff
@@ -0,0 +1,89 @@
+bv.Destroy();
+```
+
+> Note: `Destroy` deletes all data and metadata related to the broadcast variable. Use this with caution; once a broadcast variable has been destroyed, it cannot be used again.
````
```suggestion
> Note: `Destroy` deletes all data and metadata related to the broadcast variable. Use this with caution- once a broadcast variable has been destroyed, it cannot be used again.
```

bamurtaugh commented on docs/broadcast-guide.md:
```diff
@@ -0,0 +1,89 @@
+
+#### Caveat of using Destroy
+
+One important thing to keep in mind while using Broadcast variables in UDFs is to limit the scope of the variable to only the UDF that is referencing it. The [guide to using UDFs](udf-guide.md) describes this phenomenon in detail. This is especially crucial when calling `Destroy` on the broadcast variable, as if the broadcast variable that has been destroyed is visible to/accessible from other UD ...
```
```suggestion
One important thing to keep in mind while using broadcast variables in UDFs is to limit the scope of the variable to only the UDF that is referencing it. The [guide to using UDFs](udf-guide.md) describes this phenomenon in detail. This is especially crucial when calling `Destroy` on the broadcast variable. If the broadcast variable that has been destroyed is visible to or accessible from other UDFs, it gets picked up for serialization by all those UDFs, even if it is not being referenced by them. This will throw an error as .NET for Apache Spark is not able to serialize the destroyed broadcast variable.
```

bamurtaugh commented on docs/broadcast-guide.md:
```diff
@@ -0,0 +1,89 @@
+
+// Calling udf2 throws the following (unexpected) exception:
+// [Error] [JvmBridge] org.apache.spark.SparkException: Task not serializable
+df.Select(udf2(df["_1"])).Show();
```
Just to make sure I'm understanding this correctly, `udf2` throws an exception because `bv` was visible to it? Even though `udf2` doesn't reference `bv` or `udf1`?

Niharikadutta replied:
Yes that is correct, because `bv` is visible to `udf2` and is part of the shared closure of both UDFs it gets picked for serialization and because it has been destroyed, throws an error.

bamurtaugh replied:
Okay great! Your explanation and this example are awesome because that definitely made sense to me, and I am new to this topic.

bamurtaugh commented on docs/broadcast-guide.md:
````diff
@@ -0,0 +1,89 @@
+
+// Calling udf2 works fine as expected
+df.Select(udf2(df["_1"])).Show();
+```
````
This guide looks great- I didn't know what broadcast variables were, and now I feel like I actually understand them :) Do you want to end this doc with a little conclusion/wrap up section, or a link to a guide that would make sense for users to read over next? i.e. maybe mention the udf guide again, or link to a coding example or article that uses broadcast.

Niharikadutta replied:
Yes I was thinking about that, I like all the suggestions you gave, I can conclude with a summary and then link for an example to use broadcast variables, what do you think?

bamurtaugh replied:
That sounds great! :)

bamurtaugh commented on docs/udf-guide.md:
```diff
@@ -0,0 +1,159 @@
+# Guide to User-Defined Functions (UDF)
```
nit

```suggestion
# Guide to User-Defined Functions (UDFs)
```

bamurtaugh commented on docs/udf-guide.md:
```diff
@@ -0,0 +1,159 @@
+
+## What are UDFs
+
+[User-Defined Functions (UDFs)](https://spark.apache.org/docs/latest/api/java/org/apache/spark/sql/expressions/UserDefinedFunction.html) are a feature of Spark that allow developers to use custom functions to extend the vocabulary of Spark SQL’s Domain Specific Language. They transform values from a single row within a table to produce a single corresponding output value per row based on the logi ...
```
What is Spark SQL's Domain Specific Language? Could we provide a hyperlink and/or explanation to give some more context?

Niharikadutta replied:
Yeah on retrospect I think that is a jargon that doesn't need to be thrown at the user, modifying the sentence to make it more readable.

bamurtaugh commented on docs/udf-guide.md:
```diff
@@ -0,0 +1,159 @@
+
+[User-Defined Functions (UDFs)](https://spark.apache.org/docs/latest/api/java/org/apache/spark/sql/expressions/UserDefinedFunction.html) are a feature of Spark that allow developers to use custom functions to extend the vocabulary of Spark SQL’s Domain Specific Language. They transform values from a single row within a table to produce a single corresponding output value per row based on the logi ...
+
+Let's take the following as an example for a UDF definition:
```
I think it'd be great to add some more explanation here (some of which could be before or after the code example), i.e.:

* What does this UDF do?
* Are there multiple ways to define/register UDFs?
* Why do we use "Column" on the left but "string" on the right?
* Some info that the first "string" is the type of the input and the second "string" is the type of the output

Niharikadutta replied:
Have added some explanation and link to relevant sections in our code for easier understanding. Please let me know if this looks fine, thanks!

bamurtaugh replied:
Looks great, minor comment :)

bamurtaugh commented on docs/udf-guide.md:
```diff
@@ -0,0 +1,159 @@
+
+## Good to know while implementing UDFs
+
+One behavior to be aware of while implementing UDFs in .NET for Apache Spark is the way target of the UDF gets serialized. .NET for Apache Spark uses .NET Core which does not support serializing delegates, so it is instead done by using reflection to serialize the target where the delegate is defined. When multiple delegates are defined in a common scope, they have a shared closure that becomes t ...
```
```suggestion
One behavior to be aware of while implementing UDFs in .NET for Apache Spark is how the target of the UDF gets serialized. .NET for Apache Spark uses .NET Core, which does not support serializing delegates, so it is instead done by using reflection to serialize the target where the delegate is defined. When multiple delegates are defined in a common scope, they have a shared closure that becomes the target of reflection for serialization. Let's take an example to illustrate what that means.
```

bamurtaugh replied:
Could we explain delegates, reflection, and shared closure?

Niharikadutta replied:
Is it better to provide links to official docs for these terms or do we want a 1-2 line description in this doc?

bamurtaugh commented on docs/udf-guide.md:
```diff
@@ -0,0 +1,159 @@
+
+One behavior to be aware of while implementing UDFs in .NET for Apache Spark is the way target of the UDF gets serialized. .NET for Apache Spark uses .NET Core which does not support serializing delegates, so it is instead done by using reflection to serialize the target where the delegate is defined. When multiple delegates are defined in a common scope, they have a shared closure that becomes t ...
+
+Sample user code:
```
Rather than "Sample user code," perhaps we could have a sentence or two explaining what the following user code does.

Niharikadutta replied:
Done, thanks!

bamurtaugh commented on docs/udf-guide.md:
````diff
@@ -0,0 +1,159 @@
+}
+```
+As can be seen in the above IL code, both `func` and `func2` share the same closure `<>c__DisplayClass0_0`, which is the target that is serialized when serializing the delegates `func` and `func2`. Hence, even though `Func<string, string> a` is only referencing `s1`, `s2` also gets serialized when sending over the bytes to the workers.
+This can lead to some unexpected behaviors at runtime (for example in the case of using Broadcast variables, as explained in more detail in [this guide](broadcast-guide.md)), which is why we recommend restricting the visibility of the variables used in a function to that function's scope.
````
```suggestion

This can lead to some unexpected behaviors at runtime (like in the case of using [broadcast variables](broadcast-guide.md)), which is why we recommend restricting the visibility of the variables used in a function to that function's scope.
```

bamurtaugh replied:
Added extra space before sentence so that it starts as a new paragraph (if you think that'd be okay).

bamurtaugh commented on docs/udf-guide.md:
````diff
@@ -0,0 +1,159 @@
+
+Recommended user code:
+
+```csharp
````
Is this the same recommended user code as above? If so, rather than repeating it (which will make the doc longer and could be a bit confusing), I think it'd make sense just to state that you're using the same user code to accomplish task X, as was shown above.

Niharikadutta replied:
No actually this is not supposed to be the same code (my bad), thanks for pointing that out! Will make the change.

bamurtaugh commented on docs/udf-guide.md:
````diff
@@ -0,0 +1,159 @@
+}
+```
+
+Here we see that `func` and `func2` no longer share a closure and have their own separate closures `<>c__DisplayClass0_0` and `<>c__DisplayClass0_1` respectively, which when used as target for serialization mean that no more than the referenced variables get serialized for the delegate.
````
```suggestion
Here we see that `func` and `func2` no longer share a closure and have their own separate closures `<>c__DisplayClass0_0` and `<>c__DisplayClass0_1` respectively. When used as the target for serialization, nothing other than the referenced variables will get serialized for the delegate.
```

bamurtaugh commented on docs/udf-guide.md:
````diff
@@ -0,0 +1,159 @@
+```
+
+Here we see that `func` and `func2` no longer share a closure and have their own separate closures `<>c__DisplayClass0_0` and `<>c__DisplayClass0_1` respectively, which when used as target for serialization mean that no more than the referenced variables get serialized for the delegate.
+
````
Perhaps include a section with a couple sentences of wrap up, and/or link to other relevant articles and coding examples?

Niharikadutta replied:
Added a summary with links to articles, please let me know if this looks ok. Thanks!

bamurtaugh replied:
Looks really good! Added minor wording suggestion.

imback82 commented on docs/udf-guide.md:
````diff
@@ -0,0 +1,159 @@
+The above C# code generates the following IL (Intermediate language) code from the compiler:
+
+```csharp
+public class C
````
Let's give a credit to sharplap.io

imback82 commented on docs/udf-guide.md:
````diff
@@ -0,0 +1,159 @@
+}
+```
+
+The above C# code generates the following IL (Intermediate language) code from the compiler:
````
Technically speaking, the following csharp code is not an IL, but C# disassembly (I think). You can select to see IL from sharplab.io.

Niharikadutta replied:
Yes you are right, will change the wording here.

imback82 commented on docs/broadcast-guide.md:
```diff
@@ -0,0 +1,92 @@
+
+// Using the broadcast variable in a UDF:
+Func<Column, Column> udf = Udf<string, string>(
+	str => $"{str}: {bv.Value()}");
```
Can we have 4-space indentations for all our examples?

bamurtaugh commented on docs/udf-guide.md:
````diff
@@ -0,0 +1,167 @@
+        str => $"{s1} {str}");
+
+```
+The above defined UDF takes a `string` as an input (in the form of a [Column](https://github.com/dotnet/spark/blob/master/src/csharp/Microsoft.Spark/Sql/Column.cs#L14) of a [Dataframe](https://github.com/dotnet/spark/blob/master/src/csharp/Microsoft.Spark/Sql/DataFrame.cs#L24)), and returns a `string` with `hello` appended infront of the input.
````
nit
```suggestion
The above defined UDF takes a `string` as an input (in the form of a [Column](https://github.com/dotnet/spark/blob/master/src/csharp/Microsoft.Spark/Sql/Column.cs#L14) of a [Dataframe](https://github.com/dotnet/spark/blob/master/src/csharp/Microsoft.Spark/Sql/DataFrame.cs#L24)), and returns a `string` with `hello` appended in front of the input.
```

bamurtaugh commented on docs/udf-guide.md:
```diff
@@ -0,0 +1,167 @@
+
+Here we see that `func` and `func2` no longer share a closure and have their own separate closures `<>c__DisplayClass0_0` and `<>c__DisplayClass0_1` respectively. When used as the target for serialization, nothing other than the referenced variables will get serialized for the delegate.
+
+This above behavior is important to keep in mind while implementing multiple UDFs in a common scope. To know more about UDFs in general please go through the following articles that explain UDFs and how to use them: [UDFs in databricks(scala)](https://docs.databricks.com/spark/latest/spark-sql/udf-scala.html), [Spark UDFs and some gotchas](https://medium.com/@achilleus/spark-udfs-we-can-use-them- ...
```
nit wording

```suggestion
This above behavior is important to keep in mind while implementing multiple UDFs in a common scope. 

To learn more about UDFs in general, please review the following articles that explain UDFs and how to use them: [UDFs in databricks(scala)](https://docs.databricks.com/spark/latest/spark-sql/udf-scala.html), [Spark UDFs and some gotchas](https://medium.com/@achilleus/spark-udfs-we-can-use-them-but-should-we-use-them-2c5a561fde6d).
```

imback82 commented on docs/udf-guide.md:
````diff
@@ -0,0 +1,168 @@
+Let's take the following as an example for a UDF definition:
+
+```csharp
+    string s1 = "hello";
````
for all examples in this PR, can you remove leading identations?

imback82 commented on docs/udf-guide.md:
```diff
@@ -0,0 +1,168 @@
+
+## Good to know while implementing UDFs
+
+One behavior to be aware of while implementing UDFs in .NET for Apache Spark is how the target of the UDF gets serialized. .NET for Apache Spark uses .NET Core, which does not support serializing delegates, so it is instead done by using reflection to serialize the target where the delegate is defined. When multiple delegates are defined in a common scope, they have a shared closure that becomes  ...
```
I think we need to explain what `target` is and the link to target in msdn doc.

imback82 replied:
I think if you are explaining serializing, you may need to explain a bit more. This sounds like we are just serializing the target. Also, please have a link to the code where the serialization happens. Maybe a section just dedicated to a serialization would help, I think.

Niharikadutta commented:
@imback82 , @bamurtaugh  Have added a section for UDF serialization and addressed all remaining comments. Please feel free to take a look and let me know your thoughts. Thanks!

Niharikadutta commented:
@imback82 Please let me know if this looks ok or if there are any changes/suggestions you have. Thanks!

Niharikadutta commented on docs/udf-guide.md:
````diff
@@ -0,0 +1,172 @@
+    }
+}
+```
+As can be seen in the above IL code, both `func` and `func2` share the same closure `<>c__DisplayClass0_0`, which is the target that is serialized when serializing the delegates `func` and `func2`. Hence, even though `Func<string, string> a` is only referencing `s1`, `s2` also gets serialized when sending over the bytes to the workers.
````
You're right, it is the decompiled C# code, will make the change

Niharikadutta commented:
@imback82 addressed remaining comments, please let me know if it looks ok. Thanks!
